compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignConnectionTypes.cpp

```diff
@@ -29,11 +31,212 @@ class AMDAIEAssignConnectionTypesPass
+/// discovered.
+LogicalResult congestionAwareAutoAssignment(Operation *parentOp,
+                                            IRRewriter &rewriter) {
+  OpBuilder::InsertionGuard g(rewriter);
+  // Get the device model.
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(parentOp);
+  std::optional<AMDAIEDevice> maybeDevice = getConfigAMDAIEDevice(targetAttr);
+  if (!maybeDevice) {
+    return parentOp->emitOpError()
+           << "has no AMDAIEDevice in the target attribute configuration.";
+  }
+  AMDAIEDeviceModel deviceModel = AMDAIE::getDeviceModel(maybeDevice.value());
+  // Get all the connection ops and sort them into two groups: those that have
+  // already been assigned a type and those that have not.
+  SmallVector<AMDAIE::ConnectionOp> preAssignedConnections;
+  SmallVector<AMDAIE::ConnectionOp> unassignedConnections;
+  parentOp->walk([&](AMDAIE::ConnectionOp connectionOp) {
+    if (connectionOp.getConnectionType().has_value()) {
+      preAssignedConnections.push_back(connectionOp);
+    } else {
+      unassignedConnections.push_back(connectionOp);
+    }
+  });
+  SmallVector<AMDAIE::ConnectionOp> connectionOps;
+  connectionOps.reserve(preAssignedConnections.size() +
+                        unassignedConnections.size());
+  connectionOps.append(preAssignedConnections.begin(),
+                       preAssignedConnections.end());
+  connectionOps.append(unassignedConnections.begin(),
+                       unassignedConnections.end());
+  // `circuitUsage` and `packetUsage` are used to track the number of
+  // connections assigned to each physical port type. `maxCircuitUsage` is used
+  // to track the maximum number of circuit connections allowed for each
+  // physical port type, i.e. the number of channels available.
+  DenseMap<PhysPortType, uint32_t> circuitUsage, packetUsage, maxCircuitUsage;
+  auto getOrInitMaxUsage = [&](const PhysPortType &b) -> uint32_t {
+    if (maxCircuitUsage.count(b)) return maxCircuitUsage[b];
+    // Cannot just use `getNumSource/DestSwitchBoxConnections` due to shimmux.
+    if (b.portType == AMDAIE::StrmSwPortType::DMA) {
+      AMDAIETileType tileType =
+          deviceModel.getTileType(b.tileLoc.col, b.tileLoc.row);
+      maxCircuitUsage[b] =
+          deviceModel.getDmaProp<uint8_t>(tileType, AMDAIEDmaProp::NumChannels);
+    } else {
+      maxCircuitUsage[b] = (b.direction == AMDAIE::DMAChannelDir::MM2S)
+                               ? deviceModel.getNumSourceSwitchBoxConnections(
+                                     b.tileLoc.col, b.tileLoc.row, b.portType)
+                               : deviceModel.getNumDestSwitchBoxConnections(
+                                     b.tileLoc.col, b.tileLoc.row, b.portType);
+    }
+    return maxCircuitUsage[b];
+  };
+
+  // Record the final connection index associated with each physical port type.
+  DenseMap<PhysPortType, uint32_t> physPortTypeToFinalConnectionIndex;
```

Comment:
```suggestion
  DenseMap<PhysPortType, uint32_t> physPortTypeToLastConnectionIndex;
```
Nit, but you could consider renaming 'final' to 'last' as I think it's slightly more clear, but this is just a suggestion.